### PR TITLE
Fix: email signup logic

### LINKED ISSUE #16288: Can't login from email link on second reset "Oops, something is not right, please request a fresh link to sign in / sign up."
#### Challenge Name

https://beta.freecodecamp.org/en/email-signin

#### Issue Description

Site no longer responds after second deletion of account.  

I received initial email link and opened account (email 1) (test account beta@urbanware.com)
I tried browsing (screenshots below) 
Only My Profile working, couldn't access challenges
Logout/Login worked, challenges accessible
I reset challenge progress, Logout and Login results in failure, can't progress
I deleted the account, Github association (first)
I setup a new account with same email and Github (email 2)
Again only My Profile worked, this time Logout/Login didn't work, challenges inaccessible
I deleted the account, Github association (second)
I setup a new account with same email and Github (email 3)
Website produces Ooops message. 
I can no longer login

#### Browser Information

* Browser Name, Version:   Chromium, Version 63.0.3239.84 (Official Build) 
* Operating System:   Built on Ubuntu , running on LinuxMint 18.3 (64-bit)
* Mobile, Desktop, or Tablet:   Desktop (HP Elitebook 8540w Laptop)

#### Screenshot

**Initial login to beta with challenges intact.  However can't browse challenges list left menu**

001-can t finish initial challenge
002-click map at top same initial challenge error

**Logout and Login makes it work okay here** 

003-works after initial logout and login

**Reset progress, Logout and Login results in failure, can't progress, deleted account after**

006-can t progress on left menu list get two popups not one

**Second deleted account, Logout and resent email, failure, can't progress**

007-cant login again after reset

### PR DESCRIPTION
**Pull request by BerkeleyTrue**
closes #16288 
- [x] Create new AuthToken
- [x] GitHub auth should not overwrite email.

QA:
- Create new account
  - get new account message
  - get new account email
  - follow new account email link to signin
- Sign in with old account
  - get sign in message
  - get sign in email
  - follow sign in email to signin

### CODE REVIEW

File: common/models/user.json [hunk 1]
```diff
@@ -22,6 +22,10 @@
     "emailVerifyTTL": {
       "type": "date"
     },
+    "emailVerified": {
+      "type": "bolean",
```
Comment (raisedadead) [inline common/models/user.json]:
```diff
-      "type": "bolean",
+      "type": "boolean",
```

**Review by raisedadead — CHANGES_REQUESTED** [review]

**Comment by raisedadead** [comment]
Hi @BerkeleyTrue this is an excellent direction IMHO, especially moving the logic to the dedicated boot script.

I know this is super busy time for you, but could you share any of the error logs that you saw in production (or analysis of it?).

I suspect and fear that the accessToken Objects and the User Objects were not correctly linked? Just curious and concerned to find the root cause.

File: common/models/user.json (hunk 1 above)
Comment (BerkeleyTrue) [reply common/models/user.json]: fixed

**Comment by BerkeleyTrue** [comment]
@raisedadead I suspect the issue lies with some code running currently. There where also a great many db calls being made. The errors I where seeing where `cannot set headers after they are set`, which is express' way of telling us that the request has been fulfilled but some async process is attempting to the send more data. 

I didn't grab screen caps of these errors but I'd seen them enough to know that there was at the minimum an async issue.

**Review by raisedadead — APPROVED** [review]

**Comment by raisedadead** [comment]
> I suspect the issue lies with some code running currently. There where also a great many db calls being made.

Yup, the initial implementation relied on nested calls to db. I admit that it was way too procedural.

I don't remember now, but I guess the flexibility was for accounts which may not have had the `emailVerified` property set (for example, old accounts, GitHub created accounts, etc).

But, now we are enforcing only one path to create/login via email (which is good). Hence this approach suits better.

I'll definitely want to keep in mind though of the behaviour, if User Objects missing this `emailVerified` property.

Will test when this is ready, thanks for the refactor.

**Comment by BerkeleyTrue** [comment]
This is almost ready. I'd like to make one more change, adding a new model based of the accessToken called authToken. This ensures that there is a complete separation between the two types of tokens

**Comment by raisedadead** [comment]
Nice work @BerkeleyTrue !

>  I'd like to make one more change, adding a new model based of the accessToken called authToken. This ensures that there is a complete separation between the two types of tokens

Sounds reasonable to me, but with that we need to define the lifecycle methods on that too.

What I mean is we would have to implement the:
```
this.createAccessToken$({ ttl: 15 * 60 * 1000 });
```

and its associated lifecycle methods?
If not a priority we should defer it to later?

**Comment by raisedadead** [comment]
WIP - QA

Testing locally
QA:
- [x] Create new account
  - [x] get new account message
  - [x] get new account email
  - [x] follow new account email link to signin
- [ ] Sign in with old account
  - [ ] get sign in message
  - [ ] get sign in email
  - [ ] follow sign in email to signin

Observations:

- The TTL for the AccessToken Object (long term) is `NaN`
```
{
    "_id" : "PwtRSaZPbcvver5t9E0gks7xASZLq4o2znvYAqB903lSiRbbdR4wo8oUU6DUyIhu",
    "ttl" : NaN,
    "created" : ISODate("2017-12-30T07:36:48.123Z"),
    "userId" : ObjectId("5a47420bf3f3e8298faaa18f")
}
```

**Comment by raisedadead** [comment]
Additional Notes (to be taken in new PR, since they are not bugs per say because of changes):

-  The update email path breaks with a `Error: Authorization Required`
  -  Email is sent and received correctly.
  -  Clicking on the email is hitting LoopBack's native implementation (`/confirm` endpoint) and needs to be rather implemented now.

**Comment by raisedadead** [comment]
@BerkeleyTrue 

Everything looks good to me.

One thing to handle though is the destruction of the old token as I am able to do repeated logins after sign out for second time on using the same previous Login Requested email.

Steps:
1. Create new account / Logout
2. Sign in with old account (from step 1) / Logout
3. Click the same (from step 2) / Logout
4. Repeat steps 2-3.

**Comment by BerkeleyTrue** [comment]
@raisedadead Sorry, I left three commits on my local. It should fix those problems.

**Comment by raisedadead** [comment]
Thanks @BerkeleyTrue, Happy new year. I'll QA tonight and merge.

**Comment by raisedadead** [comment]
#### :sparkles: :sparkles: **WOOT!** :sparkles: :sparkles: 

Thanks @BerkeleyTrue, Nice work! Everything works like a charm. 
Lets get this on beta, and test the response times in the wild.

**Pull request merged**